src/core/etl/src/Flow/ETL/Row/Entry/MapEntry.php

```diff
@@ -0,0 +1,99 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Flow\ETL\Row\Entry;
+
+use Flow\ETL\Exception\InvalidArgumentException;
+use Flow\ETL\Row\Entry;
+use Flow\ETL\Row\Reference;
+use Flow\ETL\Row\Schema\Definition;
+
+/**
+ * @implements Entry<Entry, array{name: string, key: Entry, value: Entry}>
+ */
+final class MapEntry implements \Stringable, Entry
+{
+    use EntryRef;
+
+    /**
+     * @throws InvalidArgumentException
+     */
+    public function __construct(
+        private readonly string $name,
+        private readonly Entry $key,
+        private readonly Entry $value
+    ) {
+        if ('' === $name) {
+            throw InvalidArgumentException::because('Entry name cannot be empty');
+        }
+
+        if (!\in_array($key::class, [DateTimeEntry::class, FloatEntry::class, IntegerEntry::class, StringEntry::class, UuidEntry::class], true)) {
+            throw InvalidArgumentException::because('Entry key must be type of: DateTimeEntry, FloatEntry, IntegerEntry, StringEntry, UuidEntry, given ' . $key::class);
+        }
+    }
+
+    public function __serialize() : array
+    {
+        return ['name' => $this->name, 'key' => $this->key, 'value' => $this->value];
+    }
+
+    public function __toString() : string
+    {
+        return $this->toString();
+    }
+
+    public function __unserialize(array $data) : void
+    {
+        $this->name = $data['name'];
+        $this->key = $data['key'];
+        $this->value = $data['value'];
+    }
+
+    public function definition() : Definition
+    {
+        return Definition::map_entry($this->name);
+    }
+
+    public function is(string|Reference $name) : bool
+    {
+        if ($name instanceof Reference) {
+            return $this->name === $name->name();
+        }
+
+        return $this->name === $name;
+    }
+
+    public function isEqual(Entry $entry) : bool
+    {
+        return $this->is($entry->name()) && $entry instanceof self && $this->key->isEqual($entry->key) && $this->value->isEqual($entry->value);
+    }
+
+    public function map(callable $mapper) : Entry
+    {
+        return new self($this->name, $mapper($this->key), $mapper($this->value));
+    }
+
+    public function name() : string
+    {
+        return $this->name;
+    }
+
+    public function rename(string $name) : Entry
+    {
+        return new self($name, $this->key, $this->value);
+    }
+
+    public function toString() : string
+    {
+        return \json_encode(['key' => $this->key->toString(), 'value' => $this->value->toString()], \JSON_THROW_ON_ERROR);
+    }
+
+    /**
+     * @return string[]
+     */
+    public function value() : array
+    {
+        return [(string) $this->key->value(), (string) $this->value->value()];
```

Comment:
you can't cast key->value() to string because then numeric maps will become string maps, casting needs to be done before only for DateTime/Uuid